### ISSUE TITLE
Handling parallel request errors with Faraday adapter?

### ISSUE THREAD

**Issue by f1sherman** [HoneyryderChuck/httpx#8]
Hello and thank you for this amazing library!

I've been working on integrating it with our existing Faraday adapter and have run into some trouble trying to intelligently handle errors. In particular, there seems to be many scenarios where `httpx` will raise an error, including whenever a 4xx or 5xx response is received. When this happens it's not possible (as far as I can tell) to understand which request caused the error, and what the results of the other requests were.

```ruby
conn.in_parallel do
  response1 = conn.get('/one')
  response2 = conn.get('/two')
end # Error is raised here, with no connection to the request/response that generated it

response1.body # we never get here
response2.body
```

In our use case, depending on which requests result in errors, we would have different behavior upstream.

Would it make sense to offer an option where concurrent requests wouldn't raise errors but, instead, always return a `Response` Object with the ability to inspect the response to determine the result? Typhoeus' API seems to support this for example, where you can e.g. call `response.timed_out?` to determine if there was a timeout, rather than having an error thrown.

In any case, hopefully this made at least a slight bit of sense. Thanks for reading!

[Faraday Parallel docs for reference](https://github.com/lostisland/faraday/wiki/Parallel-requests)

**Comment by HoneyryderChuck** [comment]
Hi @f1sherman thank you for the report.

I believe this has been reported [here](https://gitlab.com/os85/httpx/-/issues/224) and being fixed [here](https://gitlab.com/os85/httpx/-/merge_requests/232) (was it you?) .

**Comment by f1sherman** [comment]
@HoneyryderChuck that wasn't me, but that's awesome, thank you!

Are there any other cases where `httpx` might raise an exception, like for a timeout?

**Comment by HoneyryderChuck** [comment]
`httpx` does not raise exceptions by itself; instead it provides `.raise_for_status` API for you to control whether to raise (see [the docs](for more info)).

In the faraday adapter, that's a different matter. The adapter for `httpx` is following the protocol for single requests, but apparently not for parallel; most because there is no protocol, just a single "reference implementation" (the typhoeus adapter), and the error handling seems to be tightly coupled to it.

**Comment by f1sherman** [comment]
Ok sounds good, thanks for all the info! I'll go ahead and close this now.

**Comment by HoneyryderChuck** [comment]
released v0.22.3

**Comment by f1sherman** [comment]
Excellent, thank you!
